### ISSUE TITLE
Not getting access to USB device

### ISSUE THREAD

**Issue by Fipps1** [scubajorgen/TomTomWatch#64]
Whenever i start the java app with the watch connected to any USB port, i get the error

ERROR: USB Error: USB error 3: Unable to claim interface: Access denied (insufficient permissions)

repeating permanently.

I'm on a Mac running Monterey 12.7.6, Java 8 Update 421

Also tried entering the command to start the app with sudo. Sadly no effort.

That’s where i'm stuck.

**Comment by Fipps1** [comment]
Solved! Created some dummy-driver and now it works.

**Comment by scubajorgen** [comment]
Thanks for solving. I don't support apple but if you have some text on how to get it working I can add it to the FAQ

Regards

Jörgen 

⁣Get BlueMail for Android ​

**Comment by axgdcode** [comment]
@Fipps1 I work on macOS, and I'm very interested in your driver 😄
Could you help me, please? 🙏

**Comment by scubajorgen** [comment]
Hi

Unfortunately I cannot offer support for MacOS since I do not have apple. 
People reported this problem more often and also solved it, but no one has given me the solution so far so I can put it in the FAQ
So if you find a solution please let me know. 

You might browse through the closed issues and contact the people who solved it

Regards 

Jörgen 

⁣Get BlueMail for Android ​

**Comment by axgdcode** [comment]
I ran the program locally, and encountered the following error:
ERROR: Error connecting: Native library not found in classpath: /org/usb4java/darwin-aarch64/libusb4java.dylib.

To address this, I followed the README instructions from the [libusb4java repository](https://github.com/usb4java/libusb4java) to build the libusb4java.dylib library. While I successfully generated the libusb4java.dylib, I have been unable to build a .jar file on my MacBook M3.

I also tried this from [yageek](https://github.com/yageek) : https://github.com/marian-m12l/studio/issues/244#issuecomment-1138296165 

an issue is opened in https://github.com/usb4java/libusb4java/issues/17

**Comment by axgd-code** [comment]
Hi everyone,

Following up on my [previous comment](https://github.com/scubajorgen/TomTomWatch/issues/64#issuecomment-2616766507) where I was stuck building the `darwin-aarch64` JAR — I finally managed to get TomTomWatch fully working on macOS Apple Silicon (M3, macOS 15 Sequoia) using [GitHub Copilot agent mode ("VibeCode")](https://code.visualstudio.com/docs/copilot/copilot-vscode-features) in VS Code.

Here is what was needed:

### 1. Build the missing `darwin-aarch64` native library

The root cause is a missing JAR: `libusb4java-1.3.0-darwin-aarch64.jar`. The upstream build fails on Apple Silicon because `src/CMakeLists.txt` doesn't link the required macOS frameworks. I've opened a PR to fix this: [usb4java/libusb4java#18](https://github.com/usb4java/libusb4java/pull/18).

Build steps (macOS ARM):
```bash
brew install libusb autoconf automake libtool cmake maven
git clone https://github.com/usb4java/libusb4java.git /tmp/libusb4java
cd /tmp/libusb4java
# Apply the fix from PR #18 (adds CoreFoundation/IOKit/Security/objc frameworks)
cd dists/darwin && chmod +x build && ./build
mvn install:install-file -Dfile=target/libusb4java-1.3.0-darwin-aarch64.jar \
  -DgroupId=org.usb4java -DartifactId=libusb4java \
  -Dversion=1.3.0 -Dclassifier=darwin-aarch64 -Dpackaging=jar
```

### 2. Run with `sudo` (required on macOS)

On macOS, libusb needs to detach the IOKit kernel driver to claim the USB interface. This requires root:
```bash
sudo java -jar TomTomWatch.jar
```

### My fork with all fixes applied

If you just want to build and run directly: https://github.com/axgd-code/TomTomWatch

It includes:
- The `darwin-aarch64` native library dependency added to `pom.xml`
- A user-friendly dialog when USB access is denied (explaining the `sudo` workaround)
- OpenStreetMap tiles switched to HTTPS (HTTP no longer works)
- `tomtomwatch.properties` found automatically next to the JAR

Confirmed working: watch connects, activities download, GPX export works 🎉

Hope this helps other macOS users!
